docs/adr/0008-freeing-objects-and-resources.md

```diff
@@ -0,0 +1,159 @@
+
+## Releasing References
+When a Rust object is  passed to the foreign side, the foreign code stores reference to a Rust `Arc<>`.
+Conversely, foreign objects passed into Rust (eg, callbacks or trait implementations) cause Rust to hold a reference to the foreign object.
+
+UniFFI needs to ensure that when the generated objects are destroyed, the reference on the other side of the FFI is released.
+It is acceptable for this release to be delayed and/or non-deterministic as long as it eventually happens.
+In particular, it is acceptable for it to be handled by garbage collection.
+
+## Closing Resources
+
+When an object holds a significant resource (eg, some scarce operating-system owned resource), it may not be acceptable to couple the resource to the lifetime of the foreign object.
+In this case, it MUST be to possible close the resource while the foreign object is still alive and before its reference is released.
+
+These same considerations apply in both directions across the FFI.
+If a garbage-collected foreign object stores a resource, then there MUST be a way to close the resource before the garbage collection happens.
+If a Rust object holds a resource, there MUST be a way to close that resource even if there may be other `Arc<>` references.
+
+## Object containers
+
+The discussion above has focused on individual objects, but in any non-trivial application many objects interact.
+For example, consider a simple `Enum` or `dictionary` - in the usual case, these objects hold references to a non-resource holding objects.
+Given the discussion above, it's acceptable for these references to be released using the standard memory management of the foreign language.
+
+However, the `Enum`, `dictionary`, `Vec`, `HashMap`, etc. may contain an object that itself contains a resource.
+While the component author is able to offer methods to close the resource on the objects themselves, they are not able to introduce methods to all the container objects which might hold the object.
+So the question here is whether UniFFI itself should try and offer such capabilities.
+
+## Foreign "sugar" for closing resources.
+
+Objects which hold significant resources offer the capability to closing those resources even while the object remains alive.
+For example, most foreign languages which expose a "file" object tend to provide a `close()` method that ensures the handle is released to the operating system.
+
+For almost all languages it is possible use something like a `try/finally` pattern to ensure the object is closed as soon as practical.
+This pattern is so common that some languages offer syntactic sugar for this express purpose.
+For example, Python offers a "context manager", Kotlin offers the `use` extension for objects that implement `Closable`, etc.
+
+Ideally, UniFFI would offer a way so consumers can opt-in to implementing this capability.
```

Comment:
I really like your separation of references and resources, and having UniFFI release references makes a lot of sense to me! But I'm not sure I follow the conclusion that UniFFI needs to concern itself with resources, and was wondering if we could talk more about that.

If a UniFFI consumer wants to define an `interface` type `T` that holds a resource—like a file or a database connection—they could manually define a `close` method, and have the Rust implementation of `T` hold a `Mutex<Option<Resource>>` that the implementation of `T::close()` method would `take()` to close the resource early. They could also `impl Drop for T`, which would take care of closing the resource once the last `Arc<T>` reference was released.

It's definitely true that, without an opt-in sugar, the UniFFI-generated Kotlin binding for `T` couldn't implement `Closeable` or `AutoCloseable`—`T` would just have a plain, vanilla `close()` method. (And, similarly, the Python binding would have a regular `close` method, and wouldn't be a context manager). But I wonder if that's okay?

I'm also wondering what this opt-in sugar would look like for languages like Swift (and maybe Ruby?) that have different idioms for resources. For example, Swift's `deinit` is the closest, but ARC doesn't guarantee precise lifetimes, and Swift seems to prefer the "`with`-prefixed function that takes a closure" idiom for resources ([`withLock`](https://developer.apple.com/documentation/os/osallocatedunfairlock/4061628-withlock), [`withExtendedLifetime`](https://developer.apple.com/documentation/swift/withextendedlifetime(_:_:)-31nv4), [`withUnsafe{Mutable}{Pointer, Bytes}`](https://developer.apple.com/documentation/swift/manual-memory-management)).

Sorry that was so wordy! I guess I'm wondering—would the opt-in sugar for resources offer enough benefits as a UniFFI feature, than having consumers handle resource management themselves?

Reply:
Great question.  Right now, I fall on the side of probably implementing `Closeable` and/or `AutoClosable` on Kotlin, maybe implementing context managers on Python, and that's it.  I'm not sure what others feel like or even if that's where I'll end up once we actually test it out.

I think the line that says "The exact mechanics for this are beyond the scope..." was supposed to communicate that we don't really have a final decision here, however the other line does seem to imply that we want to do *something*.  I softened that a bit to just say, we'll investigate it and decide later.